MazerPlatformer/GameObject.cs

```diff
@@ -162,57 +172,62 @@ public Component AddComponent(Component.ComponentType type, object value, string
             return component;
         }
 
-        private bool UpdateComponent(object newValue, Component found)
-        {
-            if (found == null) return false;
-            var oldValue = found.Value;
-            found.Value = newValue;
+        private Either<IFailure, object> UpdateComponent(object newValue, Component found)
+            => found == null
+                ? new NotFound($"Component not found to set value to {newValue}").ToFailure<Unit>()
+                : EnsureWithReturn(() =>
+                {
+                    var oldValue = found.Value;
+                    found.Value = newValue;
```

Comment:
The problem with this is that we're updating/modifying a mutable object instead of making a copy of the object, updating it and then retuning that.